### PR TITLE
Update EOL versions in `release/6.0`

### PR DESCRIPTION
**Pull request by richlander**

### CODE REVIEW

**Comment by richlander** [comment]
Related to https://github.com/dotnet/arcade/issues/8643

**Comment by richlander** [comment]
We need to update this Ubuntu 14.04 reference.

https://github.com/dotnet/arcade/blob/a4591b251c2b1d59e06af0523275ca6d4bf3aa61/azure-pipelines.yml#L16

File: eng/common/templates/jobs/source-build.yml [hunk 1]
```diff
@@ -14,7 +14,7 @@ parameters:
   # This is the default platform provided by Arcade, intended for use by a managed-only repo.
   defaultManagedPlatform:
     name: 'Managed'
-    container: 'mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7-3e800f1-20190501005343'
+    container: 'mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0'
```
Comment (MichaelSimons) [inline eng/common/templates/jobs/source-build.yml:17]:
```suggestion
    container: 'mcr.microsoft.com/dotnet-buildtools/prereqs:centos-stream9'
```

File: eng/common/templates-official/jobs/source-build.yml [hunk 2]
```diff
@@ -14,7 +14,7 @@ parameters:
   # This is the default platform provided by Arcade, intended for use by a managed-only repo.
   defaultManagedPlatform:
     name: 'Managed'
-    container: 'mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7-3e800f1-20190501005343'
+    container: 'mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7'
```
Comment (MichaelSimons) [inline eng/common/templates-official/jobs/source-build.yml:17]:
```suggestion
    container: 'mcr.microsoft.com/dotnet-buildtools/prereqs:centos-stream9'
```

**Comment by richlander** [comment]
Getting close now. Seems like we have one issue that is this:

```bash
                 Std. Error: 
[09:06:40] dbug: Executing command: 'D:\h\w\AC09092B\p\microsoft.dotnet.xharness.cli\6.0.0-prerelease.24154.4\runtimes\any\native\adb\windows\adb.exe  devices -l'
[09:06:40] fail: Error: listing attached devices / emulators: . Check that any emulators have been started, and attached device(s) are connected via USB, powered-on, unlocked and authorized.
[09:06:40] fail: Exception thrown while trying to find compatible device with architecture 'arm64-v8a'
                 System.Exception: One or more attached Android devices are offline or without USB debug permission
                    at Microsoft.DotNet.XHarness.Android.AdbRunner.GetAttachedDevicesWithProperties(String property) in /_/src/Microsoft.DotNet.XHarness.Android/AdbRunner.cs:line 426
                    at Microsoft.DotNet.XHarness.Android.AdbRunner.GetAllDevicesToUse(ILogger logger, IEnumerable`1 apkRequiredProperties, String propertyName) in /_/src/Microsoft.DotNet.XHarness.Android/AdbRunner.cs:line 467
[09:06:40] crit: Failed to find compatible device: arm64-v8a
                 Microsoft.DotNet.XHarness.Common.CLI.NoDeviceFoundException: Failed to find compatible device: arm64-v8a
                    at Microsoft.DotNet.XHarness.CLI.Commands.Android.AndroidInstallCommand.InvokeHelper(ILogger logger, String apkPackageName, String appPackagePath, IEnumerable`1 apkRequiredArchitecture, String deviceId, TimeSpan bootTimeoutSeconds, AdbRunner runner) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/Android/AndroidInstallCommand.cs:line 99
                    at Microsoft.DotNet.XHarness.CLI.Commands.Android.AndroidTestCommand.InvokeInternal(ILogger logger) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/Android/AndroidTestCommand.cs:line 73
XHarness exit code: 85 (ADB_DEVICE_ENUMERATION_FAILURE)
User command ended with 85
```

File: azure-pipelines-pr.yml [hunk 3]
```diff
@@ -18,7 +18,7 @@ variables:
 resources:
   containers:
   - container: LinuxContainer
-    image: mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-14.04-cross-0cd4667-20170319080304
+    image: mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7
```
Comment (MichaelSimons) [inline azure-pipelines-pr.yml:21]: Why centos-7 versus something newer?

Comment (richlander) [reply azure-pipelines-pr.yml:21]: Because we need to avoid breaking compat for .NET 6. CentOS 7 is part of our support contract.

https://github.com/dotnet/core/blob/main/release-notes/6.0/supported-os.md#libc-compatibility

Comment (MichaelSimons) [reply azure-pipelines-pr.yml:21]: Then why change the source-build legs off of centos-7? Additionally CentOS 7 is EOL in June isn't it?  What is the plan for the last 5-6 months?

Comment (richlander) [reply azure-pipelines-pr.yml:21]: I updated the versions to make them more conservative/consistent.

Plan: status-quo.

**Review by MichaelSimons — APPROVED** [review]
SB changes LGTM

File: eng/common/templates-official/job/source-index-stage1.yml [hunk 4]
```diff
@@ -32,10 +32,10 @@ jobs:
     - ${{ preStep }}
 
   - task: UseDotNet@2
-    displayName: Use .NET Core sdk 3.1
+    displayName: Use .NET 8 SDK
     inputs:
       packageType: sdk
-      version: 3.1.x
+      version: 8.0.x
```
Comment (chcosta) [inline eng/common/templates-official/job/source-index-stage1.yml:38]: Any reason the 6.0 sdk wouldn't work here?  ie, it simplifies things if all the places are using the same sdk where possible...

``` YAML
      - task: UseDotNet@2
        displayName: Install Correct .NET Version
        inputs:
          useGlobalJson: true
```

Comment (richlander) [reply eng/common/templates-official/job/source-index-stage1.yml:38]: It's because (I believe) the global tool has been updated to .NET 8.

File: tests/XHarness.Android.SimulatorTests.proj [hunk 5]
```diff
@@ -11,7 +11,7 @@
     <XHarnessAndroidProject Include="$(MSBuildThisFileDirectory)XHarness/XHarness.TestApks.proj">
       <AdditionalProperties>XHarnessTestX86=true;XHarnessTestX86_64=true</AdditionalProperties>
     </XHarnessAndroidProject>
-    <HelixTargetQueue Include="ubuntu.1804.amd64.android.29.open"/>
+    <HelixTargetQueue Include="ubuntu.2204.amd64.android.29.open"/>
```
Comment (premun) [inline tests/XHarness.Android.SimulatorTests.proj:14]: The build failures you're seeing are connected to this change - these are Android emulator tests. I am honestly not sure what's the story with 22.04 and emulators but you could technically leave this at `1804` for now. These are just tests that have more coverage in `dotnet/runtime` anyway.

Comment (premun) [reply tests/XHarness.Android.SimulatorTests.proj:14]: I suspect the 32 bit Android emulators are not working well in 22.04. I believe we are not testing on those anymore in runtime anyway.

So alternatively, you can keep 22.04 but remove `XHarnessTestX86=true` above.

Also we should probably use `.svc` version of the queues.

Comment (richlander) [reply tests/XHarness.Android.SimulatorTests.proj:14]:
```suggestion
    <HelixTargetQueue Include="ubuntu.1804.amd64.android.29.open"/>
```

Comment (richlander) [reply tests/XHarness.Android.SimulatorTests.proj]: @jonathanpeppers @Redth -- Do you have insight on Android requirements with ,NET 6?

Comment (jonathanpeppers) [reply tests/XHarness.Android.SimulatorTests.proj]: I think Ubuntu 22.04 has been working for us with Android emulators:

* https://github.com/xamarin/xamarin-android/blob/270c745b66e2b5a9a9bceff1ebd73cc5b44ad8fb/build-tools/automation/yaml-templates/variables.yaml#L38-L39

But we don't officially support Linux as a product for MAUI, we just keep the Android workload generally working in our spare time.

Comment (jonathanpeppers) [reply tests/XHarness.Android.SimulatorTests.proj]: We support building Android apps on Windows & Mac officially. But there might be other reasons for runtime tests to run on Linux: I've heard Linux CI machines run faster, for example.

Comment (richlander) [reply tests/XHarness.Android.SimulatorTests.proj]: I see. I misunderstood your comment a little.

Can you help me understand why Android needs to be tested in Arcade? Why isn't that something specific to runtime and maui repos?

Comment (premun) [reply tests/XHarness.Android.SimulatorTests.proj]: I can give some context on why we have this in Arcade.

Arcade contains Helix SDK which helps us run tests in other repos. Subset of Helix SDK is the XHarness SDK - this allows (mostly runtime) repos to point to a test project and the XHarness SDK builds an apk/app bundle, sends it to Helix, runs the tests on devices, collects results, increases reliability, cleans up after, etc. Basically XHarness SDK is a layer on top of our Apple/Android infrastructure so that as a whole, we provide quite reliable "testing on phones-as-a-service".

The tests that were failing here give us coverage for this functionality as runtime depends on this heavily.
The set of target platforms in these tests probably diverged from what we run in the end repos and is not in any way binding but it covers the basic matrix of simulator/device/OS.

MAUI repos do not use Helix. They use the XHarness tool directly so they don't have this layer.

Comment (richlander) [reply tests/XHarness.Android.SimulatorTests.proj]: I propose we just switch back to 1804 for .NET 6. It isn't worth the trouble. We should make sure main and net8 are good.

Comment (richlander) [reply tests/XHarness.Android.SimulatorTests.proj]: I switched the XHarness test back  to 1804. It seems like XHarness is still failing. Anyone got a guess on that?

https://github.com/dotnet/arcade/pull/14621/commits/8a0f7a477b3681f36fa5d520bd91ca4b8e7e44ba

Comment (premun) [reply tests/XHarness.Android.SimulatorTests.proj]: I believe this is just some inherent flakiness of the 6.0 world where the real world phone infra just swept from under the 6.0 branches as literally all of the parts there got updated. So I think re-running the legs a couple of times should get you a green check eventually. I realize this is not ideal but there's just not enough funding to keep the 6.0 world up to date with the everchanging MacOS and Linux versions.

Comment (premun) [reply tests/XHarness.Android.SimulatorTests.proj]: We can also merge over this, I don't think it's a big deal honestly.

**Comment by richlander** [comment]
Looks like the PR is now green. The approvals have expired.

@MichaelSimons @premun

**Review by premun — APPROVED** [review]

File: eng/common/templates/job/source-index-stage1.yml [hunk 6]
```diff
@@ -1,6 +1,6 @@
 parameters:
   runAsPublic: false
-  sourceIndexPackageVersion: 1.0.1-20210614.1
+  sourceIndexPackageVersion: 1.0.1-20240320.1
```
Comment (MichaelSimons) [inline eng/common/templates/job/source-index-stage1.yml:3]: Why was this changed?

**Review by MichaelSimons — COMMENTED** [review]
Source-build changes LGTM.  One question on a non-source-build change.

File: eng/common/templates/job/source-index-stage1.yml (hunk 6 above)
Comment (richlander) [reply eng/common/templates/job/source-index-stage1.yml:3]: Great question. The earlier version of the package was a 3.1 tool.

Enabled these changes: 

- https://github.com/dotnet/arcade/pull/14621/files#diff-9db1a2bc0710670aa4b8fc8741cffefefac50e436fdf599972ff754bbc61c745L38
- https://github.com/dotnet/arcade/pull/14621/files#diff-e908527a99e3b88dbe091e1bd3e1791b32c4281941c8bbfa45e2a13d6c02bd9bL30

Comment (premun) [reply eng/common/templates/job/source-index-stage1.yml:3]: Apologies, I thought I am the only who can merge this so I did

Comment (richlander) [reply eng/common/templates/job/source-index-stage1.yml:3]: I don't think you did anything wrong. Michael asked a question and I believe I provided a satisfactory answer. I would say that the source index package version is odd and approximately zero people are paying attention to that. It should be on a checklist for stuff we watch for.

Comment (MichaelSimons) [reply eng/common/templates/job/source-index-stage1.yml:3]: Thanks for the explanation Rich.  FYI - I reverted these change in the main PR.  https://github.com/dotnet/arcade/pull/14615/commits/832b3fc573c95e1e29940b317871e32fba652058 before merging as they appeared unrelated and I didn't get a response.

Comment (richlander) [reply eng/common/templates/job/source-index-stage1.yml:3]: Got it. I did that so that they'd match. It doesn't make much sense for the 6.0 branch to have a newer package than main.

We're missing a policy on when this package should be upgraded. The versioning scheme for the package doesn't help much. The forcing function of upgrading the package for this branch was more obvious.

@hoyosjs -- can you offer insight here.
